### PR TITLE
Fixed issue in python cmake to update wheel package

### PR DESCRIPTION
**Pull request by suryasidd**
**Description**: Fixed an issue in onnxruntime python cmake, which updates the wheel package with the providers shared library when the shared library is updated

**Motivation and Context**
- When we build the wheel package for the first time, the libonnxruntime_providers_openvino.so shared library is copied to the appropriate location but if any updates are done to the library and we build again with --build_wheel, the library is not copied, hence the wheel package is not updated with the latest library.
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Review by snnn — CHANGES_REQUESTED** [review]
The list should be the same as the list in onnxruntime.cmake. And I think the library 'onnxruntime_providers_shared' is not supposed to be linked through command line. It should be used by dlopen/LoadLibrary?

**Comment by suryasidd** [comment]
> The list should be the same as the list in onnxruntime.cmake. And I think the library 'onnxruntime_providers_shared' is not supposed to be linked through command line. It should be used by dlopen/LoadLibrary?
> 
> @RyanUnderhill Could you please help clarify?

Maybe this is not the right way to do it? Inorder for the POST_BUILD copy to be done, onnxruntime_pybind11_state this target needs to be rebuilt if libonnxruntime_providers_openvino.so is updated. So should we add this as a dependency or how do we suggest we do this ?

**Comment by RyanUnderhill** [comment]
> > The list should be the same as the list in onnxruntime.cmake. And I think the library 'onnxruntime_providers_shared' is not supposed to be linked through command line. It should be used by dlopen/LoadLibrary?
> > @RyanUnderhill Could you please help clarify?
> 
> Maybe this is not the right way to do it? Inorder for the POST_BUILD copy to be done, onnxruntime_pybind11_state this target needs to be rebuilt if libonnxruntime_providers_openvino.so is updated. So should we add this as a dependency or how do we suggest we do this ?

Yeah, my concern is that adding it to onnxruntime_pybind11_state_libs means that the binary that is built will have a static dependency on this library when the binary has no such dependency, it's loaded by our code with dlopen/LoadLibrary.

Isn't it more correct to use add_dependency of all of the shared providers + the shared provider library? This way if any of those get rebuilt, it will remake the python package. Unfortunately I'm not sure what the right way to do this in cmake is.

I'm also surprised that adding this one library makes it get rebuilt if other shared providers get changed, since the shared providers aren't dependent on the python bindings, and the python bindings aren't in the static library list also.

**Comment by jywu-msft** [comment]
> > The list should be the same as the list in onnxruntime.cmake. And I think the library 'onnxruntime_providers_shared' is not supposed to be linked through command line. It should be used by dlopen/LoadLibrary?
> > @RyanUnderhill Could you please help clarify?
> 
> Maybe this is not the right way to do it? Inorder for the POST_BUILD copy to be done, onnxruntime_pybind11_state this target needs to be rebuilt if libonnxruntime_providers_openvino.so is updated. So should we add this as a dependency or how do we suggest we do this ?

that's a good point. maybe it should something like a post build step of libonnxruntime_providers_openvino ?  (and only if python is enabled)

> > > The list should be the same as the list in onnxruntime.cmake. And I think the library 'onnxruntime_providers_shared' is not supposed to be linked through command line. It should be used by dlopen/LoadLibrary?
> > > @RyanUnderhill Could you please help clarify?
> > 
> > 
> > Maybe this is not the right way to do it? Inorder for the POST_BUILD copy to be done, onnxruntime_pybind11_state this target needs to be rebuilt if libonnxruntime_providers_openvino.so is updated. So should we add this as a dependency or how do we suggest we do this ?
> 
> Yeah, my concern is that adding it to onnxruntime_pybind11_state_libs means that the binary that is built will have a static dependency on this library when the binary has no such dependency, it's loaded by our code with dlopen/LoadLibrary.
> 
> Isn't it more correct to use add_dependency of all of the shared providers + the shared provider library? This way if any of those get rebuilt, it will remake the python package. Unfortunately I'm not sure what the right way to do this in cmake is.
> 
> I'm also surprised that adding this one library makes it get rebuilt if other shared providers get changed, since the shared providers aren't dependent on the python bindings, and the python bindings aren't in the static library list also.

i wonder if it could also be addressed as a post build step of libonnxruntime_providers_openvino ?  (and only if python is enabled)
this needs to be handled for all EP's that are built as shared libs (TensorRT, DNNL also)

**Comment by MaajidKhan** [comment]
Added a post build step for libonnxruntime_providers_openvino that copies the updated libonnxruntime_providers_openvino.so file to /onnxruntime/capi directory every time this target is rebuilt. This works perfectly fine but this post build step gets triggered every time the target is rebuilt irrespective of python is enabled or not.

**Comment by MaajidKhan** [comment]
@jywu-msft @RyanUnderhill  will this solution work for us?

File: cmake/onnxruntime_providers.cmake [hunk 1]
```diff
@@ -650,6 +650,17 @@ if (onnxruntime_USE_OPENVINO)
           ARCHIVE  DESTINATION ${CMAKE_INSTALL_LIBDIR}
           LIBRARY  DESTINATION ${CMAKE_INSTALL_LIBDIR}
           RUNTIME  DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+  if(NOT WIN32)
+    if(EXISTS ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime/capi/)
```
Comment (jywu-msft) [inline cmake/onnxruntime_providers.cmake:655]: would it be better to check whether python bindings are enabled?

File: cmake/onnxruntime_python.cmake [hunk 2]
```diff
@@ -355,18 +355,6 @@ if (onnxruntime_USE_TENSORRT)
   )
 endif()
 
-if (onnxruntime_USE_OPENVINO)
-  if(NOT WIN32)
-    add_custom_command(
```
Comment (jywu-msft) [inline cmake/onnxruntime_python.cmake:360]: does this break python bindings for windows build if you remove this?
seems like the post build logic is only for non WIN32

File: cmake/onnxruntime_providers.cmake (hunk 1 above)
Comment (MaajidKhan) [reply cmake/onnxruntime_providers.cmake]: Hi @jywu-msft . I have reverted my changes. I found the perfect fix for this issue. please, have a look at the changes now.

File: cmake/onnxruntime_python.cmake (hunk 2 above)
Comment (MaajidKhan) [reply cmake/onnxruntime_python.cmake:360]: Thanks for pointing out to this. I realized the .dll files were never getting copied to /onnxruntime/capi directory for windows builds even in the first buid and no python sample was working with OpenVINO-EP on windows.

I have made the changes now and this does the thing for us on windows and works on Linux as well.

Earlier, we were using the wrong destination for .so files to get copied to on Linux. so, the post_build step wasn't working right for Linux.

**Comment by MaajidKhan** [comment]
@jywu-msft Issue is fixed for windows and Linux. verified from my end.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**